vllm/engine/async_llm_engine.py

```diff
@@ -296,6 +360,68 @@ async def step_async(
 
         return request_outputs
 
+    def _has_remaining_steps(
+        self, seq_group_metadata_list: Optional[List[SequenceGroupMetadata]]
+    ) -> bool:
+        if not self.scheduler_config.is_multi_step:
+            return False
+
+        if seq_group_metadata_list is None:
+            return False
+
+        if not seq_group_metadata_list:
+            return False
+
+        # TODO(will) this is a sanity check for nowto make sure that all the
+        # seqs are on the same steps. Eventually we will want to do some sort of
+        # dynamic scheduling when doing multi-step decoding.
+        ref_remaining_steps = seq_group_metadata_list[0].state.remaining_steps
+        if any([
+                seq_group.state.remaining_steps != ref_remaining_steps
+                for seq_group in seq_group_metadata_list[1:]
+        ]):
+            raise AssertionError(("All running sequence groups should "
+                                  "have the same remaining steps."))
+
+        if any(seq_group.state.remaining_steps > 0
+               for seq_group in seq_group_metadata_list):
+            return True
```

Comment:
nit: a check against `ref_remaining_steps` is sufficient ?

Reply:
nice catch

Reply:
[672c0ea](https://github.com/vllm-project/vllm/pull/7000/commits/672c0ea7d1f21bd5eebd732c2874b62373efdd20)